lang/java/avro/src/main/java/org/apache/avro/Conversions.java

```diff
@@ -77,11 +78,20 @@ public String getLogicalTypeName() {
 
     @Override
     public BigDecimal fromCharSequence(CharSequence value, Schema schema, LogicalType type) {
-      return new BigDecimal(value.toString());
+      BigDecimal result = new BigDecimal(value.toString());
+      int scale = ((LogicalTypes.Decimal) type).getScale();
+      if (result.scale() != scale) {
```

Comment:
👍 
Looks good. One could allow scale increases this was as well ( <= instead of != ), but this is less magical.